### PR TITLE
lightning: local backend use external engine part 1

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/45719

Problem Summary:

### What is changed and how it works?

extract common.Engine interface, in order to let local pebble engine and external engine implement it, and let local backend can share different engine implementations.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/local/local_test.go [hunk 1]
```diff
@@ -2142,3 +2144,81 @@ func TestCtxCancelIsIgnored(t *testing.T) {
 	err := l.doImport(ctx, e, initRanges, int64(config.SplitRegionSize), int64(config.SplitRegionKeys))
 	require.ErrorContains(t, err, "the remaining storage capacity of TiKV")
 }
+
+func TestExternalEngineLoadIngestData(t *testing.T) {
```
Comment (lance6716) [inline br/pkg/lightning/backend/local/local_test.go:2148]: this is the test for new code

File: br/pkg/lightning/backend/external/engine.go [hunk 2]
```diff
@@ -35,6 +38,108 @@ type Engine struct {
...
+		ts:                 ts,
+		importedKVSize:     atomic.NewInt64(0),
+		importedKVCount:    atomic.NewInt64(0),
+	}
+}
+
+// LoadIngestData loads the data from the external storage to memory in [start,
+// end) range, so local backend can ingest it. The used byte slice of ingest data
+// are allocated from Engine.bufPool and must be released by
+// MemoryIngestData.Finish(). For external.Engine, GetIngestData must be called
```
Comment (tangenta) [inline br/pkg/lightning/backend/external/engine.go:84]:
```suggestion
// MemoryIngestData.Finish(). For external.Engine, LoadIngestData must be called
```

File: br/pkg/lightning/backend/external/engine.go [hunk 3]
```diff
@@ -35,6 +38,108 @@ type Engine struct {
...
+	memBuf := e.bufPool.NewBuffer()
+
+	if e.iter == nil {
+		iter, err := e.createMergeIter(ctx, start)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		e.iter = iter
+	} else {
+		// there should be a key that just exceeds the end key in last GetIngestData
```
Comment (tangenta) [inline br/pkg/lightning/backend/external/engine.go:104]:
```suggestion
		// there should be a key that just exceeds the end key in last LoadIngestData
```

File: br/pkg/lightning/backend/external/engine.go [hunk 4]
```diff
@@ -257,6 +374,7 @@ func (m *MemoryIngestData) GetTS() uint64 {
 
 // Finish implements IngestData.Finish.
 func (m *MemoryIngestData) Finish(totalBytes, totalCount int64) {
-	//TODO implement me
-	panic("implement me")
+	m.importedKVSize.Add(totalBytes)
+	m.importedKVCount.Add(totalCount)
+	m.memBuf.Destroy()
```
Comment (tangenta) [inline br/pkg/lightning/backend/external/engine.go:379]: Is there a situation where "MemoryIngestData.Finish()" is called concurrently?

Comment (lance6716) [reply br/pkg/lightning/backend/external/engine.go:379]: Currently one regionJob should be called its `ingestData.Finish` once

**Review by tangenta — APPROVED** [review]

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**